examples/rl/deep_q_network_cartpole.py

```diff
@@ -0,0 +1,211 @@
+
+        # use epsilon-greedy for exploration
+        if epsilon > np.random.rand(1)[0]:
+            # take random action
+            action = np.random.choice(num_actions)
+
+        # decay probability of taking random action
+        epsilon *= epsilon_decay
+        epsilon = max(epsilon_min, epsilon)
+
+        # Apply the sampled action in our environment
+        state_next, reward, done, _ = env.step(action)
+
+        episode_reward += reward
+
+        # save actions and states in replay buffer
+        action_history.append(action)
+        state_history.append(state)
+        state_next_history.append(state_next)
+        done_history.append(done)
+        rewards_history.append(reward if not done else -1)
+        state = state_next
+
+        if len(done_history) > batch_size:
+            # sample from history for replay
+            indices = np.random.choice(range(len(state_history)), size=batch_size)
+
+            state_sample = np.take(state_history, indices, axis=0)
+            state_next_sample = np.take(state_next_history, indices, axis=0)
+            rewards_sample = np.take(rewards_history, indices, axis=0)
+            done_sample = np.take(done_history, indices, axis=0)
+            action_sample = np.take(action_history, indices, axis=0)
+
+            history = zip(
+                state_sample,
+                state_next_sample,
+                rewards_sample,
+                done_sample,
+                action_sample,
+            )
+
+            updated_q_values = []
+            for (
+                sampled_state,
+                sampled_state_next,
+                sampled_reward,
+                sampled_done,
+                sampled_action,
+            ) in history:
+
+                q_update = sampled_reward
+                if not sampled_done:
+                    # Calculate the new q value
+                    sampled_state_next = tf.convert_to_tensor(sampled_state_next)
+                    sampled_state_next = tf.expand_dims(sampled_state_next, 0)
+                    # q value = reward + discount factor * expected future reward
+                    q_update = sampled_reward + gamma * np.amax(
+                        model(sampled_state_next)[0]
+                    )
+
+                sampled_state = tf.convert_to_tensor(sampled_state)
+                sampled_state = tf.expand_dims(sampled_state, 0)
+
+                # replace old q value with the updated value
+                q_value = np.array(model(sampled_state))[0]
+                q_value[sampled_action] = q_update
+
+                updated_q_values.append(q_value)
+
+            # train the model on the states and updated q values
+            model.fit(
+                tf.convert_to_tensor(state_sample),
+                tf.convert_to_tensor(updated_q_values),
+                verbose=0,
+            )
+
+            # limit the state and reward history
+            mem = 100000
```

Comment:
Presumably this parameter should be configurable at the top of the file. Also: prefer using fully-spelled out names, like "max_memory_length"